http-common/src/proxy.rs

```diff
@@ -0,0 +1,290 @@
+            }
+        };
+
+        if let Some(proxy_uri) = proxy_uri {
+            let proxy = uri_to_proxy(proxy_uri)?;
+            let proxy_connector = match identity {
+                None => hyper_proxy::ProxyConnector::from_proxy(https_connector, proxy)?,
+                Some(identity) => {
+                    // DEVNOTE: SslConnectionBuilder::build() consumes the builder. So, we need
+                    //          to create two copies of it.
+                    let proxy_tls_connector = identity_to_tls_connector(identity)?;
+
+                    let mut proxy_connector =
+                        hyper_proxy::ProxyConnector::from_proxy(https_connector, proxy)?;
+                    proxy_connector.set_tls(Some(proxy_tls_connector.build()));
+                    proxy_connector
+                }
+            };
+            Ok(MaybeProxyConnector::Proxy(proxy_connector))
+        } else {
+            Ok(MaybeProxyConnector::NoProxy(https_connector))
+        }
+    }
+}
+
+fn identity_to_tls_connector(
+    identity: (openssl::pkey::PKey<openssl::pkey::Private>, Vec<u8>),
+) -> io::Result<openssl::ssl::SslConnectorBuilder> {
+    let (key, certs) = identity;
+
+    let mut tls_connector = openssl::ssl::SslConnector::builder(openssl::ssl::SslMethod::tls())?;
+
+    let mut device_id_certs = openssl::x509::X509::stack_from_pem(&certs)?.into_iter();
+    let client_cert = device_id_certs
+        .next()
+        .ok_or_else(|| io::Error::new(io::ErrorKind::Other, "device identity cert not found"))?;
+
+    tls_connector.set_certificate(&client_cert)?;
+
+    for cert in device_id_certs {
+        tls_connector.add_extra_chain_cert(cert.clone())?;
+    }
+
+    tls_connector.set_private_key(&key)?;
+
+    Ok(tls_connector)
+}
+
+impl<C> hyper::service::Service<http::uri::Uri> for MaybeProxyConnector<C>
+where
+    C: hyper::service::Service<http::uri::Uri> + Send + Unpin + Clone + 'static,
+    C::Response:
+        AsyncRead + AsyncWrite + hyper::client::connect::Connection + Send + Unpin + 'static,
+    C::Future: Send + 'static,
+    C::Error: Into<Box<dyn std::error::Error + Sync + Send>>,
+{
+    type Response = MaybeProxyStream<C::Response>;
+    type Error = io::Error;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        let poll_status = match self {
+            MaybeProxyConnector::NoProxy(https_connector) => https_connector
+                .poll_ready(cx)
+                .map_err(|e| io::Error::new(io::ErrorKind::Other, e)),
+            MaybeProxyConnector::Proxy(proxy_connector) => proxy_connector.poll_ready(cx),
+        };
+
+        match poll_status {
+            Poll::Ready(Ok(())) => Poll::Ready(Ok(())),
+            Poll::Ready(Err(e)) => Poll::Ready(Err(io::Error::new(io::ErrorKind::Other, e))),
```

Comment:
`e` is already a `std::io::Error`, so you don't need this entire `match`. Just let the first one return its value without binding it.

Reply:
Fixed.